Add guide section about platform-dependent nature of jextract (#237)
openjdk/jextract: doc/GUIDE.md

being generated. (All of these are discussed in more detail in the
[Using The Code Generated By Jextract section](#using-the-code-generated-by-jextract)).

> Generally speaking, the code that jextract generates is dependent on the platform on which

Reviewer: ```suggestion
Generally speaking, the bindings generated by jextract depend on the platform on which
```

Reviewer: The above is to streamline things a bit (optional)

jextract is running. For example, when a C header file is processed by the C pre-processor,
it is possible for code in the header file to detect the current platform using
pre-processor directives, and expand to a different result based on that platform (see
also the section on [pre-processor definitions](#preprocessor-definitions)). Additionally,
different built in C types can have different formats depending on the platform (for

> example, the `long` type has different formats on Linux and Windows). Both of these are

Reviewer: Maybe we also want to mention padding in structs?

Author: That only really comes into play on ABIs that use a different packing (like the AIX ABI), so it doesn't really seem that important to mention. Any way, the things I mentioned are just a couple of examples. With pre-processor directives a header could really create any kind of platform dependency it wants.

examples of things that can lead jextract to generate different outputs depending on the
platform on which it is run. Care should be taken when generating bindings with jextract
on one platform, and using those bindings on another platform, as the header files may
contain platform-dependent code that can cause the generated code to misbehave when used
on another platform.

However, it is also possible for a C library to be written in such a way that it is not

> platform dependent: a so-called _portable_ library. Sharing the bindings generated for a

Reviewer: Perhaps here we might say what moves portable libraries might do - such as use `long long` instead of `long`, or using sized numeric types e.g. `int32`.

Author: Added this: 

> These libraries, for instance, use data types that have the same format on all supported platforms (such as `long long` instead of `long`, or an explicitly-sized integer type such as `int64_t`).

portable library across different platforms should work without issues. It is typically

> advisable to generate different sets of bindings, once on each platform on which the

Reviewer: ```suggestion
advisable to generate different sets of bindings, one on each platform on which the
```
